public/app/features/trails/services/sorting.ts

```diff
@@ -0,0 +1,137 @@
+
+import { reportExploreMetrics } from '../interactions';
+
+import { getLabelValueFromDataFrame } from './levels';
+
+export const sortSeries = memoize(
+  (series: DataFrame[], sortBy: string, direction = 'asc') => {
+    if (sortBy === 'alphabetical') {
+      return sortSeriesByName(series, 'asc');
+    }
+
+    if (sortBy === 'alphabetical-reversed') {
+      return sortSeriesByName(series, 'desc');
+    }
+
+    if (sortBy === 'outliers') {
+      initOutlierDetector(series);
+    }
+
+    const reducer = (dataFrame: DataFrame) => {
+      try {
+        if (sortBy === 'outliers') {
+          return calculateOutlierValue(series, dataFrame);
+        }
+      } catch (e) {
+        console.error(e);
+        // ML sorting panicked, fallback to stdDev
+        sortBy = ReducerID.stdDev;
+      }
+      const fieldReducer = fieldReducers.get(sortBy);
+      const value =
+        fieldReducer.reduce?.(dataFrame.fields[1], true, true) ?? doStandardCalcs(dataFrame.fields[1], true, true);
+      return value[sortBy] ?? 0;
+    };
+
+    const seriesCalcs = series.map((dataFrame) => ({
+      value: reducer(dataFrame),
+      dataFrame: dataFrame,
+    }));
+
+    seriesCalcs.sort((a, b) => {
+      if (a.value !== undefined && b.value !== undefined) {
+        return b.value - a.value;
+      }
+      return 0;
+    });
+
+    if (direction === 'asc') {
+      seriesCalcs.reverse();
+    }
+
+    return seriesCalcs.map(({ dataFrame }) => dataFrame);
+  },
+  (series: DataFrame[], sortBy: string, direction = 'asc') => {
+    const firstTimestamp = series.length > 0 ? series[0].fields[0].values[0] : 0;
+    const lastTimestamp =
+      series.length > 0
+        ? series[series.length - 1].fields[0].values[series[series.length - 1].fields[0].values.length - 1]
+        : 0;
+    const firstValue = series.length > 0 ? getLabelValueFromDataFrame(series[0]) : '';
+    const lastValue = series.length > 0 ? getLabelValueFromDataFrame(series[series.length - 1]) : '';
+    const key = `${firstValue}_${lastValue}_${firstTimestamp}_${lastTimestamp}_${series.length}_${sortBy}_${direction}`;
+    return key;
+  }
+);
+
+const initOutlierDetector = (series: DataFrame[]) => {
+  if (!wasmSupported()) {
+    return;
+  }
+
+  // Combine all frames into one by joining on time.
+  const joined = outerJoinDataFrames({ frames: series });
+  if (!joined) {
+    return;
+  }
+
+  // Get number fields: these are our series.
+  const joinedSeries = joined.fields.filter((f) => f.type === FieldType.number);
+  const nTimestamps = joinedSeries[0].values.length;
+  const points = new Float64Array(joinedSeries.flatMap((series) => series.values as number[]));
+
+  try {
+    const detector = OutlierDetector.dbscan({ sensitivity: 0.4 }).preprocess(points, nTimestamps);
+    outliers = detector.detect();
+  } catch (e) {
+    console.error(e);
+  }
+};
+
+let outliers: OutlierOutput | undefined = undefined;
```

Comment:
From what I can tell this is the same way that it's done in Explore Logs but I'm a bit confused about why we're using a global for `outliers` here rather than just a local var passed to `initOutlierDetector` / `calculateOutlierValue`?

It also looks like that could cause a stale value of `outliers` which could crash if:

1. The user selects outlier sorting and it works correctly (`outliers` will not be `undefined`)
2. They switch to another view
3. They select outlier sorting again and it fails, so that `outliers` is never overwritten but is still not `undefined`
4. `calculateOutlierValue` doesn't know that `outliers` is stale, and tries to index the old value of `outliers.seriesResults` with an invalid value of `index`

Perhaps I'm missing something though!

cc @matyax

Reply:
You make a great point here @sd2k . I would say it shouldn't be a problem because if it panics and then throws, it should be catched by the try/catch on line 25, which falls back to stddev.